File: dispatch/ios/CLAID/Sources/CLAID/RemoteFunction/RemoteFunctionRunnable.swift
//
//  RemoteFunctionRunnable.swift
//  CLAID
//
//  Created by Patrick Langer on 23.02.2025.
//

public final class RemoteFunctionRunnable<ReturnType: Sendable, each Parameter: Sendable> : AbstractRemoteFunctionRunnable {
    private let functionName: String
    private let function: @Sendable (repeat each Parameter) -> ReturnType
    private let mutatorHelpers: [AbstractMutator]
    private let params: (repeat each Parameter)
    
    init(functionName: String,
         returnTypeExample: ReturnType,
         paramExamples: repeat each Parameter,
         function: @escaping @Sendable (repeat each Parameter) -> ReturnType) {
        
        self.functionName = functionName
        self.function = function
         
        var helpers: [AbstractMutator] = []
        repeat helpers.append(TypeMapping.getMutator(type(of: each paramExamples)))
        self.mutatorHelpers = helpers
        
        self.params = (repeat each paramExamples)
    }
    
    func executeRemoteFunctionRequest(_ rpcRequest: Claidservice_DataPackage) -> Claidservice_DataPackage? {
        let executionRequest = rpcRequest.controlVal.remoteFunctionRequest
        let remoteFunctionIdentifier = executionRequest.remoteFunctionIdentifier
        
        let parameterPayloads = executionRequest.parameterPayloads
        if parameterPayloads.count != mutatorHelpers.count {
            print("Failed to execute RemoteFunctionRunnable \(functionName). Number of parameters do not match.")
            return nil
        }
     
        var count = 0
        let parameter = (repeat unpack(each params, &count, parameterPayloads))

        
        let result: ReturnType = function(repeat each parameter)
        let remoteFunctionResult = RemoteFunctionRunnableResult.makeSuccessfulResult(result)
        return makeRPCResponsePackage(result: remoteFunctionResult, rpcRequest: rpcRequest)
    }
    
    // Helper function to bridge runtime data to static types
    private func unpackParameters(payloads: [Claidservice_Blob], _ par: (repeat each Parameter)) -> (repeat each Parameter) {
        // This is the tricky part - we need to map runtime data to the compile-time Parameter pack
        
       // let xx =  (repeat unpack(each params))
        
        /*var index = 0
        for var param in repeat each self.params {
            var stubPackage = Claidservice_DataPackage()
            stubPackage.payload = payloads[index]
            let helper = mutatorHelpers[index]
            param = repeat helper.getPackagePayloadAsObject(stubPackage) as! each Parameter
            index += 1
        }
        
        return ()
        */
        // This is a conceptual placeholder - actual casting to [(repeat each Parameter)] is complex
        return params
    }
    
    func unpack<T>(_ paramExample: T, _ index: inout Int, _ payloads: [Claidservice_Blob]) -> T {
        var stubPackage = Claidservice_DataPackage()
        
        stubPackage.payload = payloads[index]
        let helper = mutatorHelpers[index]
        let val = helper.getPackagePayloadAsObject(stubPackage) as! T
        index += 1
        return val
    }
    
    private func makeRPCResponsePackage(
            result: RemoteFunctionRunnableResult<ReturnType>,
            rpcRequest: Claidservice_DataPackage
        ) -> Claidservice_DataPackage? {
            var responsePackage = Claidservice_DataPackage()
            
            responsePackage.sourceModule = rpcRequest.targetModule
            responsePackage.targetModule = rpcRequest.sourceModule
            
            var ctrlPackage = responsePackage.controlVal
            ctrlPackage.ctrlType = .ctrlRemoteFunctionResponse
            ctrlPackage.runtime = rpcRequest.controlVal.runtime
            
            let executionRequest = rpcRequest.controlVal.remoteFunctionRequest
            ctrlPackage.remoteFunctionReturn = makeRemoteFunctionReturn(result: result, executionRequest: executionRequest)
            responsePackage.controlVal = ctrlPackage
            setReturnPackagePayload(package: &responsePackage, result: result)
            
            return responsePackage
        }
        
        private func makeRemoteFunctionReturn(
            result: RemoteFunctionRunnableResult<ReturnType>,
            executionRequest: Claidservice_RemoteFunctionRequest
        ) -> Claidservice_RemoteFunctionReturn {
            var remoteFunctionReturn = Claidservice_RemoteFunctionReturn()
            
            remoteFunctionReturn.executionStatus = result.getStatus()
            remoteFunctionReturn.remoteFunctionIdentifier = executionRequest.remoteFunctionIdentifier
            remoteFunctionReturn.remoteFutureIdentifier = executionRequest.remoteFutureIdentifier
            
            return remoteFunctionReturn
        }
        
        private func setReturnPackagePayload(
            package: inout Claidservice_DataPackage,
            result: RemoteFunctionRunnableResult<ReturnType>
        ) {
            if let returnValue = result.getReturnValue(), ReturnType.self != Void.self {
                let mutator = TypeMapping.getMutator(ReturnType.self)
                package = mutator.setPackagePayload(packet: package, value: returnValue)
            }
        }
}

File: dispatch/python/claid/remote_function/remote_function_runnable.py
from typing import Callable, TypeVar, Any, List, Tuple
import logging

from module.type_mapping.type_mapping import TypeMapping
from remote_function.remote_function_runnable_result import RemoteFunctionRunnableResult
from dispatch.proto.claidservice import *
from logger.logger import Logger

ReturnType = TypeVar('ReturnType')
Parameter = TypeVar('Parameter')

class RemoteFunctionRunnable:
    def __init__(self, function_name: str, function: Callable[..., ReturnType], return_type, *param_examples: Any):
        """
        :param function_name: Name of the function to be executed.
        :param function: A callable function with a variable number of parameters.
        :param param_examples: Example parameters to infer types.
        """
        self.function_name = function_name
        self.function = function
        self.param_examples = param_examples
        self.return_type = return_type

        # Create mutator helpers based on parameter types
        print("Param examples: ", param_examples)
        self.mutator_helpers = [TypeMapping.get_mutator(param) for param in param_examples]

    def execute_remote_function_request(self, rpc_request: DataPackage) -> DataPackage:
        execution_request = rpc_request.control_val.remote_function_request
        remote_function_identifier = execution_request.remote_function_identifier

        parameter_payloads = execution_request.parameter_payloads
        if len(parameter_payloads) != len(self.mutator_helpers):
            Logger.log_fatal(f"Failed to execute RemoteFunctionRunnable {self.function_name}. "
                          f"Number of parameters do not match. Got {len(parameter_payloads)} parameters but expected {len(self.mutator_helpers)}")
            return None

        # Unpacking payloads into parameters
        parameters = self.unpack_parameters(parameter_payloads)

        # Execute function
        print("Executing function")
        result: ReturnType = self.function(*parameters)
        print("Executed function: ", result)
        # Create a successful function result
        remote_function_result = RemoteFunctionRunnableResult.make_successful_result(result)
        return self.make_rpc_response_package(remote_function_result, rpc_request)

    def unpack_parameters(self, payloads: List[Blob]) -> Tuple[Any, ...]:
        """
        Converts raw payloads into parameter values using type mutators.
        """
        parameters = []
        for index, (helper, param_example) in enumerate(zip(self.mutator_helpers, self.param_examples)):
            stub_package = DataPackage()
            stub_package.payload = payloads[index]
            value = helper.get_package_payload(stub_package)
            parameters.append(value)
        return tuple(parameters)

    def make_rpc_response_package(self, result: RemoteFunctionRunnableResult, rpc_request: DataPackage) -> DataPackage:
        """
        Creates a response package after executing the remote function.
        """
        print("Making response package")
        response_package = DataPackage()
        response_package.source_module = rpc_request.target_module
        response_package.target_module = rpc_request.source_module

        ctrl_package = ControlPackage()
        ctrl_package.ctrl_type = CtrlType.CTRL_REMOTE_FUNCTION_RESPONSE
        ctrl_package.runtime = rpc_request.control_val.runtime

        execution_request = rpc_request.control_val.remote_function_request
        ctrl_package.remote_function_return = self.make_remote_function_return(result, execution_request)
        response_package.control_val = ctrl_package
        self.set_return_package_payload(response_package, result)
        return response_package

    def make_remote_function_return(self, result: RemoteFunctionRunnableResult, execution_request: RemoteFunctionRequest) -> RemoteFunctionReturn:
        """
        Creates a return structure for a remote function call.
        """
        remote_function_return = RemoteFunctionReturn()
        remote_function_return.execution_status = result.get_status()
        remote_function_return.remote_function_identifier = execution_request.remote_function_identifier
        remote_function_return.remote_future_identifier = execution_request.remote_future_identifier
        return remote_function_return

    def set_return_package_payload(self, package: DataPackage, result: RemoteFunctionRunnableResult):
        """
        Sets the return value in the package if applicable.
        """
        return_value = result.get_return_value()
        if return_value is not None:
            mutator = TypeMapping.get_mutator(return_value)
            mutator.set_package_payload(package, return_value)



    @classmethod
    def is_data_type_supported(clz, data_type):
        try:
            mutator = TypeMapping.get_mutator(data_type)
            return True
        except Exception as e:
            return False
